src/audio/sources/mp3.cpp

```diff
@@ -0,0 +1,194 @@
+{
+	assert(this->context != nullptr);
+
+	long rate = 0;
+	mpg123_getformat(this->context, &rate, nullptr, nullptr);
+
+	assert(0 < rate);
+	// INT32_MAX isn't a typo; if we compare against UINT32_MAX, we'll
+	// set off sign-compare errors, and the sample rate shouldn't be above
+	// INT32_MAX anyroad.
+	assert(rate <= INT32_MAX);
+	return static_cast<std::uint32_t>(rate);
+}
+
+std::uint64_t Mp3AudioSource::Seek(std::uint64_t in_samples)
+{
+	assert(this->context != nullptr);
+
+	// See BytesPerSample() for an explanation of this ChannelCount().
+	auto mono_samples = in_samples * this->ChannelCount();
+
+	// Have we tried to seek past the end of the file?
+	auto clen = static_cast<unsigned long>(mpg123_length(this->context));
+	if (clen < mono_samples) {
+		Debug() << "mp3: seek at" << mono_samples << "past EOF at"
+		        << clen << std::endl;
+		throw SeekError(MSG_SEEK_FAIL);
+	}
+
+	if (mpg123_seek(this->context, mono_samples, SEEK_SET) == MPG123_ERR) {
+		Debug() << "mp3: seek failed:" << mpg123_strerror(this->context)
+		        << std::endl;
+		throw SeekError(MSG_SEEK_FAIL);
+	}
+
+	// The actual seek position may not be the same as the requested
+	// position.
+	// mpg123_tell gives us the exact mono-samples position.
+	return mpg123_tell(this->context) / this->ChannelCount();
+}
+
+Mp3AudioSource::DecodeResult Mp3AudioSource::Decode()
+{
+	assert(this->context != nullptr);
+
+	auto buf = reinterpret_cast<unsigned char *>(&this->buffer.front());
+	size_t rbytes = 0;
+	int err = mpg123_read(this->context, buf, this->buffer.size(), &rbytes);
+
+	DecodeVector decoded;
+	DecodeState decode_state;
+
+	if (err == MPG123_DONE) {
+		Debug() << "mp3: end of file" << std::endl;
+		decode_state = DecodeState::END_OF_FILE;
+	} else if (err != MPG123_OK && err != MPG123_NEW_FORMAT) {
+		Debug() << "mp3: decode error:" << mpg123_strerror(this->context)
+		        << std::endl;
+		decode_state = DecodeState::END_OF_FILE;
+	} else {
+		decode_state = DecodeState::DECODING;
+
+		// Copy only the bit of the buffer occupied by decoded data
+		auto front = this->buffer.begin();
+		decoded = DecodeVector(front, front + rbytes);
+	}
+
+	return std::make_pair(decode_state, decoded);
+}
+
+SampleFormat Mp3AudioSource::OutputSampleFormat() const
+{
+	assert(this->context != nullptr);
+
+	int encoding = 0;
+	mpg123_getformat(this->context, nullptr, nullptr, &encoding);
+
+	switch (encoding) {
+		case MPG123_ENC_UNSIGNED_8:
+			return SampleFormat::PACKED_UNSIGNED_INT_8;
+		case MPG123_ENC_SIGNED_8:
+			return SampleFormat::PACKED_SIGNED_INT_8;
+		case MPG123_ENC_SIGNED_16:
+			return SampleFormat::PACKED_SIGNED_INT_16;
+		case MPG123_ENC_SIGNED_32:
+			return SampleFormat::PACKED_SIGNED_INT_32;
+		case MPG123_ENC_FLOAT_32:
+			return SampleFormat::PACKED_FLOAT_32;
+		default:
+			// We shouldn't get here, if the format was set up
+			// correctly earlier.
+			assert(false);
+			return SampleFormat::PACKED_UNSIGNED_INT_8;
```

Comment:
err? also, needs a NOT_REACHED() function or equivalent, rather than misusing assert
please, this isn't cccp